src/main/flight/pid.c

```diff
@@ -436,28 +439,94 @@ STATIC_UNIT_TESTED FAST_CODE_NOINLINE float calcHorizonLevelStrength(void)
+            //angle feedforward averaging at RCrate inactive for low speed links and 2-point for fast links
+            if (pidRuntime.feedforwardAveraging){
+                pidRuntime.angleFeedforward[axis] = laggedMovingAverageUpdate(&pidRuntime.angleFeedforwardAvg[axis].filter, pidRuntime.angleFeedforward[axis]);
+            }
+            pidRuntime.previousAngleTarget[axis] = angleTarget;
+            //feedforward transition
+            angleFeedforwardInput *= pidRuntime.feedforwardTransitionFactor > 0 ? MIN(1.0f, getRcDeflectionAbs(axis) * pidRuntime.feedforwardTransitionFactor) : 1.0f;
+        } else {
+            if (duplicateCount > 1) {
+                angleFeedforwardInput = 0;
+            }
+        }
+    }
+    //angle feedforward filter at the PID loop rate
+    float angleFeedforward = pt3FilterApply(&pidRuntime.angleFeedforwardPt3[axis], angleFeedforwardInput);
+    pidRuntime.angleFeedforward[axis] = angleFeedforwardInput;
+    angleFeedforward *= pidRuntime.levelFeedforwardGain;
+    pidRuntime.angleDerivative[axis] = currentAngleDerivative;
+    currentAngleDerivative *= pidRuntime.angleDerivativeGain;
 #endif
-    angle = constrainf(angle, -levelAngleLimit, levelAngleLimit);
-    const float errorAngle = angle - ((attitude.raw[axis] - angleTrim->raw[axis]) / 10.0f);
     if (FLIGHT_MODE(ANGLE_MODE) || FLIGHT_MODE(GPS_RESCUE_MODE)) {
-        // ANGLE mode - control is angle based
-        const float setpointCorrection = errorAngle * pidRuntime.levelGain;
-        currentPidSetpoint = pt3FilterApply(&pidRuntime.attitudeFilter[axis], setpointCorrection);
+        // ANGLE mode - control is angle based with P term based on angle error and feedforward based on angle setpoint velocity
+        rawSetpoint = angleErrorCorrection + angleFeedforward - currentAngleDerivative;
     } else {
         // HORIZON mode - mix of ANGLE and ACRO modes
-        // mix in errorAngle to currentPidSetpoint to add a little auto-level feel
-        const float setpointCorrection = errorAngle * pidRuntime.horizonGain * horizonLevelStrength;
-        currentPidSetpoint += pt3FilterApply(&pidRuntime.attitudeFilter[axis], setpointCorrection);
+        // mix in angleError to currentPidSetpoint to add a little auto-level feel
+        rawSetpoint += angleError * horizonLevelStrength * pidRuntime.horizonGain;
     }
-    return currentPidSetpoint;
+    //logging
+    if (axis == FD_ROLL) {
+        DEBUG_SET(DEBUG_ANGLEMODE, 0, lrintf(angleTarget * 10.0f));
+        DEBUG_SET(DEBUG_ANGLEMODE, 1, lrintf(currentAngle * 10.0f));
+        DEBUG_SET(DEBUG_ANGLEMODE, 2, lrintf(angleErrorCorrection * 10.0f));
+        DEBUG_SET(DEBUG_ANGLEMODE, 3, lrintf(-currentAngleDerivative * 10.0f));
+        DEBUG_SET(DEBUG_ANGLE_FEEDFORWARD, 0, lrintf(angleTarget * 10.0f));
+        DEBUG_SET(DEBUG_ANGLE_FEEDFORWARD, 1, lrintf(currentAngle * 10.0f));
+        DEBUG_SET(DEBUG_ANGLE_FEEDFORWARD, 2, lrintf(angleFeedforward * 10.0f));
```

Comment:
temporary debug
```suggestion
        DEBUG_SET(DEBUG_ANGLE_DERIVATIVE, 0, lrintf(currentAngle * 10.0f));
        DEBUG_SET(DEBUG_ANGLE_DERIVATIVE, 1, lrintf(currentAngleDerivative * 10.0f));
        DEBUG_SET(DEBUG_ANGLE_DERIVATIVE, 2, lrintf(rawSetpointUnsmoothed * 10.0f));
        DEBUG_SET(DEBUG_ANGLE_DERIVATIVE, 2, lrintf(rawSetpoint * 10.0f));
```

Reply:
It's better to have P and D on one debug and feedforward on another for tuning. Please update blackbox accordingly.

We could add P to the feedforward debug?

Reply:
Well, in this case, the D factor isn't just opposing P.  It's opposing wobble, and wobble comes both from P and Feedforward.  Each peaks at their own time point, depending on the filtering of FF.
We kind of need D to peak in opposition to the predominant stuff we want to remove.
My model of delayed FF makes the early contribution from error the primary source.

In any case for the first requirement we need to see the filtering of D.  So we need to visualise unsmoothed derivative and smoothed derivative to check that we are controlling the noise and see how much unfiltered noise there is.